packages/cursorless-engine/src/util/allocateHats/maxByFirstDiffering.test.ts

```diff
@@ -0,0 +1,40 @@
+import * as assert from "assert";
+import { maxByAllowingTies } from "./maxByFirstDiffering";
+
+// known good but slow
+function goldenMaxByAllowingTies<T>(arr: T[], fn: (item: T) => number): T[] {
+  const max = Math.max(...arr.map(fn));
+  return arr.filter((item) => fn(item) === max);
+}
+
+suite("maxByFirstDiffering", () => {
+  test("maxByAllowingTies", () => {
+    const testCases: number[][] = [
+      [],
+      [0],
+      [1],
+      [-Infinity],
+      [+Infinity],
+      [0, 0],
+      [0, 1],
+      [1, 0],
+      [-Infinity, -Infinity],
+      [-Infinity, +Infinity],
+      [+Infinity, -Infinity],
+      [+Infinity, 0],
+      [0, +Infinity],
+      [-Infinity, 0],
+      [0, -Infinity],
+      [0, 1, 0, 1],
+      [1, 0, 1, 0],
+      [0, 1, 1, 0],
+      [1, 0, 0, 1],
+    ];
+
+    testCases.forEach((testCase) => {
+      const actual = maxByAllowingTies(testCase, (x) => x);
+      const expected = goldenMaxByAllowingTies(testCase, (x) => x);
```

Comment:
Cute. I like it

Reply:
An old favorite approach from fuzz testing. Particularly well suited to performance optimization. :)

https://en.wikipedia.org/wiki/Differential_testing